deep_qa/common/params.py

```diff
@@ -1,52 +1,184 @@
+                    log_recursively(value, new_local_history)
+                else:
+                    logger.param(self.history + self.history + " : " + key + " : " + str(value))
 
-def get_choice_with_default(params: Dict[str, Any],
+        logger.info("Converting Params object to dict; logging of default "
+                    "values will not occur when dictionary parameters are "
+                    "used subsequently.")
+        logger.info("CURRENTLY DEFINED PARAMETERS: ")
+        log_recursively(self.params, self.history)
+        return self.params
+
+    def __check_is_dict(self, new_history, value):
+        if isinstance(value, dict):
+            new_history = new_history if self.history == "" else self.history + "." + new_history
+            return Params(value, new_history)
+        else:
+            return value
+
+    def __getitem__(self, key):
+        return self.__check_is_dict(key, self.params[key])
+
+    def __setitem__(self, key, value):
+        self.params[key] = value
+
+    def __delitem__(self, key):
+        del self.params[key]
+
+    def __iter__(self):
+        return iter(self.params)
+
+    def __len__(self):
+        return len(self.params)
+
+
+def _get_choice_error_message(value: Any, choices: List[Any], name: str=None) -> str:
+    if name:
+        return '%s not in acceptable choices for %s: %s' % (value, name, str(choices))
+    else:
+        return '%s not in acceptable choices: %s' % (value, str(choices))
+
+
+def pop_choice_with_default(params: Dict,
                             key: str,
                             choices: List[Any],
                             default: Any=None,
-                            name: str=None):
+                            name: str=None) -> Any:
     """
-    Like get_choice, but with a default value.  If `default` is None, we use the first item in
-    `choices` as the default.
+    Performs the same function as the pop_with_default_method of Params,
+    but is required in order to deal with places that the Params object is not
+    welcome, such as inside Keras layers.
     """
     try:
-        return get_choice(params, key, choices, name)
+        value = params.pop(key)
     except KeyError:
         if default is None:
-            return choices[0]
-        if default not in choices:
-            raise ConfigurationError(_get_choice_error_message(default, choices, name))
-        return default
+            value = choices[0]
+        else:
+            value = default
+    if value not in choices:
+        raise ConfigurationError(_get_choice_error_message(default, choices, name))
+
+    logger.param("UNSCOPED PARAMETER: Retrieve default arguments manually."
+                 + " : " + key + " : " + str(value))
+
+    return value
 
 
-def assert_params_empty(params: Dict[str, Any], class_name: str):
+def assert_params_empty(params: Union["Params", Dict], class_name: str):
```

Comment:
This shouldn't need quotes around `Params`, because `Params` has already been defined at this point.